ens-improvement-proposals/ensip-15-normalization-standard.md

```diff
@@ -0,0 +1,343 @@
+| **Created** | 2023-04-03                           |
+
+## Abstract
+
+This ENSIP standardizes Ethereum Name Service (ENS) name normalization process outlined in [ENSIP-1 § Name Syntax](./ensip-1-ens#name-syntax).
+
+## Motivation
+
+* Since [ENSIP-1](./ensip-1-ens) (originally [EIP-137](https://eips.ethereum.org/EIPS/eip-137)) was finalized in 2016, Unicode has [evolved](https://unicode.org/history/publicationdates.html) from version 8.0.0 to 15.0.0 and incorporated many new characters, including complex emoji sequences. 
+* ENSIP-1 does not state the version of Unicode.
+* ENSIP-1 implies but does not state an explicit flavor of IDNA processing. 
+* [UTS-46](https://unicode.org/reports/tr46/) is insufficient to normalize emoji sequences. Correct emoji processing is only possible with [UTS-51](https://www.unicode.org/reports/tr51/).
+* Validation tests are needed to ensure implementation compliance.
+* The success of ENS has encouraged spoofing via the following techniques:
+	1. Insertion of zero-width characters.
+	1. Using names which normalize differently between algorithms. 
+	1. Using names which appear differently between applications and devices.
+	1. Substitution of confusable (look-alike) characters.
+	1. Mixing incompatible scripts.
+
+## Specification
+
+### Versioning
+* Unicode version `15.0.0`
+* [spec.json](./ensip-15/spec.json) contains all necessary data for normalization.
+* [nf.json](./ensip-15/nf.json) contains all necessary data for [Unicode Normalization Forms](https://unicode.org/reports/tr15/) NFC and NFD.
+
+### Algorithm
+
+* Normalization is the process of canonicalizing a name before for hashing.
+* It is idempotent: applying normalization multiple times produces the same result.
+* For user convenience, leading and trailing whitespace should be trimmed before normalization, as all whitespace codepoints are disallowed.  Inner characters should remain unmodified.
+* No string transformations (like case-folding) should be applied.
+
+1. Split the name into [labels](./ensip-1-ens#name-syntax).
+1. [Normalize](#normalize) each label.
+1. Join the labels together into a name again. 
+1. The result is normalized and ready for [hashing](./ensip-1-ens#namehash-algorithm).
+
+### Normalize
+
+1. [Tokenize](#tokenize) — transform the label into **Text** and **Emoji** tokens.
+	* If there are no tokens, the label cannot be normalized.
+1. Apply NFC to each **Text** token.
+1. Strip `FE0F` from each **Emoji** token.
+1. [Validate](#validate) — check if the tokens are valid and obtain the **Label Type**.
+	* The **Label Type** and **Restricted** state may be presented to user for additional security.
+1. Concatenate the tokens together.
+	* Return the normalized label.
+
+### Tokenize
+
+Given a string, convert to codepoints, and produce a list of **Text** and **Emoji** tokens, each with a payload of codepoints.
+
+1. Allocate an empty codepoint buffer.
+1. Find the longest emoji sequence that matches the remaining input.

```

Comment:
@adraffy Maybe it is simpler to explain to use regular expression constructed as alternative of all emojis.

Reply:
My convention was that the initial paragraph describes the process of the following itemized steps.

Is the question how to convert a string to codepoints?  If so, can I make a general disclaimer that all strings are Unicode strings and 1:1 with sequence of codepoints?

Reply:
> @adraffy Maybe it is simpler to explain to use regular expression constructed as alternative of all emojis.

The naive regex for this is enormous (48KB) although you're correct that it's conceptually simple, since it's just: `/^(emoji1|emoji2|...)/u` where `\uFE0F` is replaced with `\uFE0F?` and `*` is replaced with `\*`

In v8, it appears to only takes 3MB of memory to store that regex (~30K DFA states).  I wonder if that's actually a better way to parse the emoji...  I'll have to do some tests.

Reply:
My question here is how do you "find the longest emoji sequence that matches the remaining input"?

Reply:
> > @adraffy Maybe it is simpler to explain to use regular expression constructed as alternative of all emojis.
> 
> The naive regex for this is enormous (48KB) although you're correct that it's conceptually simple, since it's just: `/^(emoji1|emoji2|...)/u` where `\uFE0F` is replaced with `\uFE0F?` and `*` is replaced with `\*`
> 
> In v8, it appears to only takes 3MB of memory to store that regex (~30K DFA states). I wonder if that's actually a better way to parse the emoji... I'll have to do some tests.

@adraffy It is not better for performance, in Python we have done that instead of using trie as you did.

I have proposed this here for answering @Arachnid question. Maybe regex is simpler than explaining the trie.

Reply:
It's not the "concept of longest" that is unspecified - it's how you determine the longest sequence of *emoji*. You should provide specific directions on identifying where an emoji sequence ends.

Reply:
Ah, gotcha.  I added the following near the top of the document:
> An Emoji Sequence is a [single entity composed of one of more](https://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries) emoji characters and emoji components.

And made sure I was using the word **Emoji Sequence** consistently throughout the document (except for the `spec.json` part where I say the word emoji a million times.)

I also added djstrong's regex suggestion.

Reply:
Also added definition of a string:

* A string is a sequence of Unicode codepoints.